### PR TITLE
Fix installing organs via surgery, and organ removal

### LINKED ISSUE #2295: I can't reattach lungs after I detached them
#### Description of issue
After detaching an organ, such as a lung, the game assumes that the character still has that lung and prevents it from being put back in.

#### Difference between expected and actual behavior
If I have separated the lung, I am expected to be able to put it back in later.

#### Steps to reproduce
Detach out the lung
Try to stick it back with a tool

#### Issue bingo

- [x] Issue could be reproduced at least once
- [x] Issue could be reproduced by different players
- [x] Issue could be reproduced in multiple rounds
- [x] Issue happened in a recent (less than 7 days ago) round
- [x] [Couldn't find an existing issue about this](https://github.com/NebulaSS13/Nebula/issues)

### PR DESCRIPTION
**Pull request by PsyCommando**
## Description of changes
The organ wouldn't get cleared from the mob's organ lists properly, so fixed that.
* ~~Instead of slapping in some hardcoded snowflake code for the mmi_holder I've made some procs to override the behavior of organ detach during surgery.~~ 
* Removed the contents of the vital_organs list in the species. Since removing some organs would cause unintended death and brain damage.

EDIT:
* Remade the logic behind removing/adding external organs, since it was causing some serious issues.
* Partially remade the logic behind the internal organs in order to handle surgeries better.

Fixes #2295 

## Changelog

:cl:
bugfix: Fix organ surgery.
/:cl:

### CODE REVIEW

**Comment by PsyCommando** [comment]
Okay, after messing around with dismembering and adding and removing organs, I think this is ready for review.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (MistakeNot4892) [inline code/modules/mob/living/carbon/carbon_organs.dm]: This is redundant, human get_contained_external_atoms already subtracts the external organ list, which stumps should be in.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (MistakeNot4892) [inline code/modules/mob/living/carbon/carbon_organs.dm]: This is _exactly_ why I was against moving the removal and addition procs onto the mob.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: It doesn't, because I had to put that there in order to prevent the stumps from coming out. I couldn't find another implementation.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: Its more of an helper than anything. Even before the organ changes, it was just a long snippet copy pasted a million times to do that.
Also, I'm not entirely sure what you're really getting to anyways?

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: Actually, after thinking about it, I know what's up. Stumps were getting installed on initialize before.. Going to fix that.

Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: Nah, they still get dumped. Gonna need to mess with the gib proc methinks.

Comment (MistakeNot4892) [reply code/modules/mob/living/carbon/carbon_organs.dm]:
````
/mob/living/carbon/get_contained_external_atoms()
	. = ..()
	LAZYREMOVE(., get_organs())
````
Current `dev`, `living.dm`.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (MistakeNot4892) [reply code/modules/mob/living/carbon/carbon_organs.dm]: Metastatizing new specialized procs on the mob is exactly what I was expecting to happen from your changes to scope and encapsulation of organ behavior. It's going to be more brittle, it's going to harder to override for specific organs, everything I spoke about at length prior to your human init PR being merged.

It's moot, but I felt obligated to say something. You can disregard this comment.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: > ```
> /mob/living/carbon/get_contained_external_atoms()
> 	. = ..()
> 	LAZYREMOVE(., get_organs())
> ```
> 
> Current `dev`, `living.dm`.

Yeah, realized that earlier when I fixed stumps not getting added to the external_organ list.
But the line does nothing in our particular case. Might be because the organ list are cleared before dump_contents() is called in destroy. So they still get dumped anyways. Could also be the gib proc kicking them out the body on gib. ~~But it shouldn't since stumps cannot be amputated.~~

EDIT: They can be amputated, but only for surgery.. But seems like the check to dismember only looks at whether the amputation flag is on.. Seems like the flag is being used with possibly mutually exclusive states in mind.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: I'm gonna be honest here. Its hard to be more brittle than it was already. And I actually noticed it was holding up pretty well in that regard, even seeing an improvement.

Also, I did try to integrate more of your approach from your PR, but a lot of it didn't work, or didn't take things into account. It didn't seem like you tested/investigated a lot of it before. (EX: things like mobs not having eyes for example because get_organ would be used in too many cases and override organ zones and return the head when passed BP_EYES) 
It also touched to so many different systems that I didn't feel like spending time to fill in the blanks, and deal with the massive amount of testing myself, since that's was way out of scope for me. And you should keep in mind my changes were never meant to be the be all, end all of organs. I fully expected there would be a need for further refactors as humans are overhauled. If you think you can do it better, by all means be my guest! I'm just doing what I can with what I got.

Also, concerning your comment, I integrated the surgery behavior into add_organ and remove_organ now via a boolean param. I think its less clear and less overridable than having a proc to handle it, but I guess it means less specialized procs. 
Will be pushing once I test everything.

File: code/modules/mob/living/carbon/carbon_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/carbon/carbon_organs.dm]: I added a is_droppable() proc to decide if an organ should be allowed to exist outside the mob. Otherwise it just gets deleted instead of dropped. Seems to be working well to prevent stumps and the root limb from just getting kicked out. 

Only weird thing I'm noticing is that gib() doesn't seem to run consistently. Sometimes everything gets kicked out like it should, some other times, the lower body organs don't get kicked out for some reasons. Not entirely sure what's happening there, since everything seems 100% fine structure wise. Its like gib is called in parallel with destruction almost.

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm]: Uh... No? If anything, this should be a radial menu. If a dismembered limb has several sub-parts, either branching or linear, you should be able to choose which to cut off rather than blindly hacking at it.

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm]: No "starts trying to cut" message was ever sent. That's probably a big oversight; you should be able to see if someone is hacking up a dismembered organ *before* they finish or stop doing it.

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm:358]: Are there assurances that children can't be cyclic? It might be good to check if it's already in the list, and only add it and call get_limbs_recursive if it isn't.

Comment (noelle-lavenza) [reply code/modules/organs/external/_external.dm:358]: (Also, would kind of like to see either separate functions instead of arguments like `no_stumps`, or the ability to pass predicates/procs/callbacks in. A whitelist and blacklist argument would probably suffice, so to replicate `no_stumps` you'd pass a callback to `is_stump`, or the path to is_stump, as the blacklist argument. Kind of a shame there are no inline functions in DM, but I think it's a better design than control coupling.)

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm]:
```suggestion
		if(owner?.can_autoheal(dam_type))
```

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm:1402]: I'd suggest maybe moving this to a proc on the parent, so subtypes can control additional references to remove. Something like `/obj/item/organ/external/proc/remove_child(obj/item/organ/to_remove)`?

File: code/modules/organs/organ.dm
Comment (noelle-lavenza) [inline code/modules/organs/organ.dm]: Nitpick, but technically it's "droppable".

File: code/modules/organs/organ.dm
Comment (noelle-lavenza) [inline code/modules/organs/organ.dm:554]: Is there a reason stumps aren't a subtype that we can do an istype on?

File: code/modules/species/species.dm
Comment (noelle-lavenza) [inline code/modules/species/species.dm]:
```suggestion
	if(!LAZYLEN(vital_organs))
```

File: code/modules/surgery/limb_reattach.dm
Comment (noelle-lavenza) [inline code/modules/surgery/limb_reattach.dm:151]: This will runtime, won't it? `if(existing && O != existing) CRASH("mob/living/carbon/add_organ(): '[O]' tried to overwrite [src]'s existing organ '[existing]' in slot '[O.organ_tag]'!")` Or is it fine because it's technically the same organ?

File: code/modules/surgery/organs_internal.dm
Comment (noelle-lavenza) [inline code/modules/surgery/organs_internal.dm:282]: Personally, I'd move this into something like `/obj/item/organ/proc/get_manipulate_sound()` or something, so that it can be used for attaching, detaching, and other surgery interactions that need sounds to differ based on prosthetic status and other bits.

File: code/unit_tests/organ_tests.dm
Comment (noelle-lavenza) [inline code/unit_tests/organ_tests.dm]: I don't think this works; you'd need to use a for loop if you want a list of stumps. As it stands, the body of the for loop will never run. However, since you check LAZYLEN, LAZYLEN of a single datum is always 0; therefore, the test will always pass even if stumps drop.
```suggestion
```

File: code/unit_tests/organ_tests.dm
Comment (noelle-lavenza) [inline code/unit_tests/organ_tests.dm]:
```suggestion
	for(var/obj/item/organ/external/stump/O in world)
```

**Review by noelle-lavenza — CHANGES_REQUESTED** [review]
A few complaints/concerns, but bravo for replacing those span strings with the SPAN_ defines; that had really been bothering me.

Additionally, have you tested that braindeath and decapitation are still lethal?

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Actually, I'm not sure what's going on with that. The do_after call behaves really weirdly and doesn't pop a progressbar like it should and I couldn't figure why. I tried several different ways, and tried to compare it with other calls to it elsewhere and can't figure out why it doesn't behave.. So its either a bug, or I'm doing something stupid I can't see.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Well, the behavior is consistent with the removal of internal organs that's right below. 
Also, the radial menu is kind of a bit annoying to use. But, I guess I could change it to that relatively easily.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm:358]: > Are there assurances that children can't be cyclic? It might be good to check if it's already in the list, and only add it and call get_limbs_recursive if it isn't.

Childrens have to be unique in the mob's list and organs list. If there were cyclic references, this proc would be the last of your worries, cause everything would start breaking up. (True story XD )

> (Also, would kind of like to see either separate functions instead of arguments like `no_stumps`, or the ability to pass predicates/procs/callbacks in. A whitelist and blacklist argument would probably suffice, so to replicate `no_stumps` you'd pass a callback to `is_stump`, or the path to is_stump, as the blacklist argument. Kind of a shame there are no inline functions in DM, but I think it's a better design than control coupling.)

I guess that could be a future improvement. But its a bit out of scope right now. This is mainly a bugfix PR XD
Truth be told, a lot of the human code is still really wonky and will need work, so I'm not too much in a hurry to start polishing stuff like that, and having it get changed anyways a few weeks later when someone refactors all this.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: true

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm:1402]: That's something I considered. But, its not a very common case to have special behavior when a limb is outside a mob. I don't really have any applicable use cases for those in the codebase rn, so someone would probably just delete the proc since its unused. Limb behavior outside mobs currently is basically just ceasing any processing and just being in a proper hierarchy so icon updates and stuff works.

But, yeah, its a bit out of scope for this bugfix imho. But it would make sense to add that in the future.

File: code/modules/organs/organ.dm
Comment (PsyCommando) [reply code/modules/organs/organ.dm:554]: That's actually a proc from /obj/item/organ/external that I brought down to /obj/item/organ. Needed it here to avoid doing extra type check + cast since a lot of proc work on /organ in general.

Honestly I'm not sure if doing a type check is faster than using a proc. But I'm assuming if it was there someone probably checked it before?

File: code/modules/organs/organ.dm
Comment (PsyCommando) [reply code/modules/organs/organ.dm]: Nice catch XD

File: code/modules/surgery/limb_reattach.dm
Comment (PsyCommando) [reply code/modules/surgery/limb_reattach.dm:151]: I fixed that already. It will only runtime if we try to replace a different organ in the same slot. Not if its the same one.

File: code/modules/surgery/organs_internal.dm
Comment (PsyCommando) [reply code/modules/surgery/organs_internal.dm:282]: True, but that's again out of scope. I was mostly just updating the surgery code to work properly, and I noticed that prosthetics made the squelch sound when installing them just like organic limbs, so I fixed it. 
But yeah, that's a good idea.

File: code/unit_tests/organ_tests.dm
Comment (PsyCommando) [reply code/unit_tests/organ_tests.dm]: You're right! Not sure why I thought it returned a list..

File: code/unit_tests/organ_tests.dm
Comment (PsyCommando) [reply code/unit_tests/organ_tests.dm]: (Had to commit it in two part since apparently I didn't add it to the batch properly.)

**Review by noelle-lavenza — APPROVED** [review]
All the things I found before are now fixed, so it seems good to me.

File: code/modules/organs/external/_external.dm
Comment (noelle-lavenza) [inline code/modules/organs/external/_external.dm]:
```suggestion
		all_limbs.Add(child.get_limbs_recursive(no_stumps))
```

**Review by noelle-lavenza — CHANGES_REQUESTED** [review]
Ackshually, forgot one.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Good catch!

**Review by noelle-lavenza — DISMISSED** [review]
*Now* it seems good.

File: code/modules/mob/living/living_organs.dm
Comment (MistakeNot4892) [inline code/modules/mob/living/living_organs.dm]: Is this check duplicated?

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm]: Probably should check if the organ is still attached and present in the mob.

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [reply code/modules/organs/external/_external.dm]: Radial menu or a modal selector would be preferable if the target doll isn't usable.

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [reply code/modules/organs/external/_external.dm]: Jade is referring to the absence of a 'Dave starts cutting off thingo' before all of this processing logic fires.

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm]: This is going to insert nulls if nothing is returned from the recursive limb proc.

File: code/modules/mob/living/living_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/living_organs.dm]: It is because its the same condition but it needs to happen after do_uninstall, and its too simple of a check for me to bother to place the result into a var.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: It inserts nulls if the list passed to Add is empty?

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: That's not something you can do on a mob. Its for loose limbs lying on the ground. Added a check if its in the parent limb instead.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Tested it on wrench and it does not insert nulls at all?

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Alright. Fixed that, pushing it soon.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Added a radial menu to that and internal organ removing too.

File: code/modules/mob/living/living_organs.dm
Comment (PsyCommando) [reply code/modules/mob/living/living_organs.dm]: Changed the second condition.

**Comment by PsyCommando** [comment]
All changes are in. Just waiting for approval.

File: code/modules/mob/living/living.dm
Comment (MistakeNot4892) [inline code/modules/mob/living/living.dm]: Needs to be LAZYREMOVE so that `.` is nulled if the list is cleared, to line up with previous implementation.

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm:115]: Unless the logic has been moved somewhere, `parent.children` is not being properly updated.

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm]: Please check for `TOOL_SAW` (ideally use the tool procs for the interaction).

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm]: As noted above, please either manually check for tool quality, or use `do_tool_interaction` a la
````
if(W.do_tool_interaction(TOOL_WELDER, user, src,        \
	  max(5, round((stored_max_health-stored_health) / 5)), \
	  "repairing the damage to", "repairing the damage to", \
	  "You fail to patch the damage to \the [src].",        \
	  fuel_expenditure = 1
))
````

File: code/modules/organs/external/_external.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/_external.dm:1572]: Needs to use `LAZYREMOVE`.

File: code/modules/organs/organ.dm
Comment (MistakeNot4892) [reply code/modules/organs/organ.dm:554]: Typechecks should generally be avoided. Class shouldn't be used as a categorization system.

File: code/modules/organs/external/standard.dm
Comment (MistakeNot4892) [inline code/modules/organs/external/standard.dm:42]: Will this interfere with the amputation mechanic that destroys the human mob when all limbs other than the chest are removed?

File: code/modules/organs/organ.dm
Comment (MistakeNot4892) [inline code/modules/organs/organ.dm:41]: Why are these being nulled?

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm:115]: /obj/item/organ/Destroy() calls do_uninstall when stored inside a limb. So no need to do that here.

File: code/modules/organs/external/standard.dm
Comment (PsyCommando) [reply code/modules/organs/external/standard.dm:42]: That has nothing to do with that. That mechanic is handled in its own conditional that checks if only a single limb is left then gib:

_external.dm, line 1001
```	owner.remove_organ(src, TRUE, FALSE, ignore_children)
	if(istype(last_owner) && !QDELETED(last_owner) && LAZYLEN(last_owner.get_external_organs()) <= 1)
		last_owner.physically_destroyed(FALSE, disintegrate)
```

This is solely to categorize things as droppable or not, so that when we gib the mob, or individual limbs, its not going to drop say the chest or stumps. I made a proc for it so its no hardcoded and doing type checks or expecting things to get deleted in time at the same time the mob is gibbed like it used to, since it was unreliable after testing.

File: code/modules/organs/organ.dm
Comment (PsyCommando) [reply code/modules/organs/organ.dm:41]: Just because its a reference. Even if species are singletons, I got the reflex to clear any dangling refs.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm:1572]: done

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Done. 
But. kinda uncomfortable to add in something that's barely used anywhere else in a bugfix tbh.
Also the defines for the return values are defined too late to be used in here. So I used literals instead for handling the no message case, and success case.

File: code/modules/organs/external/_external.dm
Comment (PsyCommando) [reply code/modules/organs/external/_external.dm]: Done

File: code/modules/mob/living/living.dm
Comment (PsyCommando) [reply code/modules/mob/living/living.dm]: Fixed

File: code/modules/organs/organ.dm
Comment (MistakeNot4892) [reply code/modules/organs/organ.dm:41]: It's a bit untidy, and in this case it's not a circular ref so it's unnecessary for GC. Not a huge issue.

**Review by MistakeNot4892 — APPROVED** [review]
Not sure why we need unit tests for stumps, but they're harmless at worst.

Nothing else springs out at me as wrong but I am also not signing off on the fixes working as I am (as demonstrated) largely unfamiliar with the new organ code.

**Comment by PsyCommando** [comment]
> Not sure why we need unit tests for stumps, but they're harmless at worst.

Cause people keep breaking them. And now we know when they do. Its been a reoccurring issue since at least 2016 on bay.

> Nothing else springs out at me as wrong but I am also not signing off on the fixes working as I am (as demonstrated) largely unfamiliar with the new organ code.

Did a lot of surgery tests in the last month. They work. And if something is still broken, I'll just make another PR to fix them.

Also, about the changes. Its all essentially very similar to the old system, except additon/removal of organs has been streamlined into general purpose procs instead of being re-implemented by anything using organs. Also now you can make sure installing/removing organs won't call dangerous things on the parent when you really don't want them to.

In a nutshell:

add_organ() calls:
- do_install()
- on_install_effects()

Here's what they do:
* add_organ() do installing stuff on the mob, organ lists etc.. 
* do_install() do installing stuff on the limb/organ itself, its internal state, and the limb hierarchy whether its inside a mob, or outside, without messing with the mob as much as possible. Can be called to attach loose limbs to other loose limbs as well.
* on_install_effects() handles the fluff of organ installation, like interactive elements, messages, things you wouldn't want to happen if you were say swapping 2 organs in place via code that aren't necessary for the organ to work for example. Its a separate proc because we can skip and override those steps at will.

Those 3 procs and their parameters covers most of the use cases for dealing with organs in the entire codebase.
The removal procs are similar. 
Its is a huge plus compared to having to manually dig through organs, figure out what var checks you have to do to safely retrieve the organs and change the proper lists, and in what limbs, every single times you do something slightly different with organs..

**Review by comma — APPROVED** [review]

**Pull request merged**
